File: src/main/java/de/tectoast/emolga/utils/ModManager.java
package de.tectoast.emolga.utils;

import de.tectoast.emolga.commands.Command;
import de.tectoast.jsolf.JSONObject;

import java.util.ArrayList;

public class ModManager {
    private static final ArrayList<ModManager> modManagers = new ArrayList<>();
    private final String name;
    private JSONObject dex;
    private JSONObject learnsets;
    private JSONObject moves;
    private JSONObject typechart;

    public ModManager(String name, String datapath) {
        this.name = name;
        new Thread(() -> this.dex = Command.loadSD(datapath + "pokedex.ts", Constants.DEXJSONSUB), "ModManager " + name + " Dex").start();
        new Thread(() -> this.learnsets = Command.loadSD(datapath + "learnsets.ts", Constants.LEARNSETJSONSUB), "ModManager " + name + " Learnsets").start();
        new Thread(() -> this.moves = Command.loadSD(datapath + "moves.ts", Constants.MOVESJSONSUB), "ModManager " + name + " Moves").start();
        new Thread(() -> this.typechart = Command.loadSD(datapath + "typechart.ts", Constants.TYPESJSONSUB), "ModManager " + name + " Typechart").start();
        modManagers.add(this);
    }

    public static ModManager getByName(String name) {
        return modManagers.stream().filter(m -> m.name.equals(name)).findFirst().orElse(null);
    }
    public JSONObject getDex() {
        return dex;
    }

    public JSONObject getLearnsets() {
        return learnsets;
    }

    public JSONObject getMoves() {
        return moves;
    }

    public JSONObject getTypechart() {
        return typechart;
    }

}

File: src/main/kotlin/de/tectoast/emolga/utils/ModManager.kt
package de.tectoast.emolga.utils

import de.tectoast.emolga.commands.Command
import de.tectoast.emolga.commands.Command.Companion.load
import de.tectoast.emolga.utils.json.showdown.Learnset
import de.tectoast.emolga.utils.json.showdown.Pokemon
import de.tectoast.emolga.utils.json.showdown.TypeData
import dev.minn.jda.ktx.util.SLF4J
import kotlinx.coroutines.*
import kotlinx.serialization.json.JsonObject
import org.slf4j.Logger

class ModManager(name: String, datapath: String) {
    lateinit var dex: Map<String, Pokemon>
    lateinit var learnsets: Map<String, Learnset>
    lateinit var moves: JsonObject
    lateinit var typechart: Map<String, TypeData>

    init {
        scope.launch { dex = load(datapath + "pokedex.json") }
        scope.launch { learnsets = load(datapath + "learnsets.json") }
        scope.launch { moves = load(datapath + "moves.json") }
        scope.launch { typechart = load(datapath + "typechart.json") }
        if (name == "default") default = this
        modManagers.add(this)
    }

    companion object {
        private val modManagers = ArrayList<ModManager>()
        private val logger: Logger by SLF4J
        lateinit var default: ModManager
        val scope = CoroutineScope(Dispatchers.IO + SupervisorJob() + CoroutineExceptionHandler { _, t ->
            logger.error("ERROR IN MODMANAGER SCOPE", t)
            Command.sendToMe("Error in modmanager scope, look in console")
        })
    }
}
